base/generator.jl

```diff
@@ -128,3 +128,51 @@ IteratorEltype(::Type) = HasEltype()  # HasEltype is the default
 IteratorEltype(::Type{Generator{I,T}}) where {I,T} = EltypeUnknown()
 
 IteratorEltype(::Type{Any}) = EltypeUnknown()
+
+
+abstract type IteratorIndexable end
+struct Indexable <: IteratorIndexable end
+struct HasEachIndex{I} <: IteratorIndexable end
+struct NotIndexable <: IteratorIndexable end
+
+"""
+    IteratorIndexable(itertype::Type) -> IteratorIndexable
+
+Given the type of an iterator, return one of the following values:
+
+* `NotIndexable()` if `getindex` is not implemented or the elements it returns do not match the iterator's eltype.
+* `Indexable()` if `getindex` is implemented for the iterator but `eachindex` is not.
+* `HasEachIndex{I}` if both `getindex` and `eachindex` are implemented for the iterator, where I is the type of `eachindex(iterator)`.
+
+The default value (for iterators that do not define this function) is `NotIndexable()`.
+
+```jldoctest
+julia> Base.IteratorIndexable([1,2,3])
+Base.HasEachIndex{Base.OneTo{$Int}}()
+
+julia> Base.IteratorIndexable(Ref(1))
+Base.Indexable()
+
+julia> Base.IteratorIndexable(Set([1,2,3]))
+Base.NotIndexable()
+
+julia> Base.IteratorIndexable(Dict(1=>'a', 2=>'b'))
+Base.NotIndexable()
+
+```
+"""
+IteratorIndexable(x) = IteratorIndexable(typeof(x))
+IteratorIndexable(::Type) = NotIndexable()
+IteratorIndexable(::Type{<:Ref}) = Indexable()
+IteratorIndexable(::Type{<:Tuple}) = HasEachIndex{Base.OneTo{Int}}()
+IteratorIndexable(::Type{<:Number}) = HasEachIndex{Base.OneTo{Int}}()
+IteratorIndexable(::Type{<:AbstractArray{T,1}}) where T = HasEachIndex{Base.OneTo{Int}}()
+IteratorIndexable(T::Type{<:AbstractString}) = HasEachIndex{Base.EachStringIndex{T}}()
+
+function IteratorIndexable(T::Type{<:AbstractArray{E,N}}) where {E,N}
+    if IndexStyle(T) === IndexLinear()
+        HasEachIndex{Base.OneTo{Int}}()
+    else
+        HasEachIndex{CartesianIndices{N, NTuple{N, Base.OneTo{Int}}}}()
```

Comment:
Doesn't this assume one-based indexing? What about OffsetArrays?

Reply:
It does, I tried mirroring the default values for `eachindex` for each type but that's not really good enough, hmmm.. Maybe `AbstractArray` should default to `Indexable()` and then concrete Base types can use these default values.

Reply:
I'm not sure, but can't you simply ask for `typeof(eachindex(A))`? Can't that be obtained at compile time?

Reply:
How about this?
```julia
struct UnknownEachIndex <: IteratorIndexable end
struct KnownEachIndex{I} <: IteratorIndexable end
HasEachIndex{I} = Union{UnknownEachIndex, KnownEachIndex{I}}
HasEachIndex() = UnknownEachIndex()
HasEachIndex{I}() where I = KnownEachIndex{I}()

function IteratorIndexable(x)
    I = IteratorIndexable(typeof(x))
    I isa UnknownEachIndex ? HasEachIndex{typeof(eachindex)}() : I
end

IteratorIndexable(::Type) = NotIndexable()
IteratorIndexable(::Type{<:Ref}) = Indexable()
IteratorIndexable(::Type{<:AbstractArray}) = HasEachIndex()
IteratorIndexable(::Type{<:AbstractString}) = HasEachIndex()
IteratorIndexable(::Type{<:Tuple}) = HasEachIndex{Base.OneTo{Int}}()
IteratorIndexable(::Type{<:Number}) = HasEachIndex{Base.OneTo{Int}}()
```

Reply:
Actually, there's a problem with this, in that `(UnknownEachIndex <: HasEachIndex{Int}) == true`. I suppose this works better?
```julia
abstract type HasEachIndex{I} end
struct UnknownEachIndex <: HasEachIndex{Any} end
struct KnownEachIndex{I} <: HasEachIndex{I} end
# [...]
```

Reply:
Or did you mean that I should remove the typevar and ask for the type outside the trait? I mostly did it as a shortcut because I could easily include it, but since it turned out to be more complicated than I thought, than it's not worth the hassle.

Reply:
I meant returning `HasEachIndex{typeof(eachindex(A))}`. Then you wouldn't need the branches.

Reply:
Well, I don't have `A` at the point I return the first `HasEachIndex`, I could if I semi-duplicated the methods like this:
```julia
IteratorIndexable(x) = IteratorIndexable(typeof(x))
IteratorIndexable(x::AbstractArray) = HasEachIndex{typeof(eachindex(x))}()
IteratorIndexable(x::AbstractString) = HasEachIndex{typeof(eachindex(x))}()

IteratorIndexable(::Type) = NotIndexable()
IteratorIndexable(::Type{<:Ref}) = Indexable()
IteratorIndexable(::Type{<:AbstractArray}) = HasEachIndex()
IteratorIndexable(::Type{<:AbstractString}) = HasEachIndex()
IteratorIndexable(::Type{<:Tuple}) = HasEachIndex{Base.OneTo{Int}}()
IteratorIndexable(::Type{<:Number}) = HasEachIndex{Base.OneTo{Int}}()
```
But is that necessary? I think that's an easy branch for the compiler to solve:
```julia
julia> @code_typed IteratorIndexable([1,2,3,4])
CodeInfo(
1 ─     Base.arraysize(x, 1)::Int64
└──     return $(QuoteNode(KnownEachIndex{Base.OneTo{Int64}}()))
) => KnownEachIndex{Base.OneTo{Int64}}
```

Reply:
OTOH, what you currently have is perhaps fine for a fallback method. Other array types should simply implement their own specialized trait methods.

Reply:
That's what I thought, now I am worried about breakages if e.g. we add an optimization in `Base` that depends on the default result, [like this one for `drop`](https://gist.github.com/fcard/2756d9895ba450a4f1880812b4fa47b6#file-fastdrop-jl-L34-L46), and then `drop` will be broken for any array that changes the type of `eachindex` to something else other than `AbstractRange` until they implement `IteratorIndexable`.

However, I am a bit dissatisfied with this new one too because it will make `IteratorIndexable(x) != IteratorIndexable(typeof(x))` for some types, which seems weird because it's `==` for all the other ones, although maybe `IteratorIndexable(x)` could be seen as a separate function that adds instance information to the trait. Or maybe we should properly separate it into another generic function so `IteratorIndexable(x) == IteratorIndexable(typeof(x))` but there is a `iteratorindexable(x)` (or something) function that expands the trait with instance information.

Hmmm..

Reply:
I've thought and decided that the type of `eachindex(itr)` doesn't belong in the trait for now. Instead I've added a sister function `eachindextype`:
```julia
eachindextype(x) = eachindextype(x, IteratorIndexable(x))
eachindextype(x, ::HasEachIndex) = typeof(eachindex(x))
eachindextype(x, ::IteratorIndexable) = missing
```
It's not a trait, since traits are meant to be based on the type of the argument (as far as I know), and the concrete return type of a function is a runtime property of instances. (It's why we can't just use `Base.return_types`, other than it being a private part of the inference engine)

Reply:
New issue: `eachindextype` would be only be defined for `x` if `IteratorIndexable(x) == HasEachIndex()`, but that means that `eachindextype(Dict(1=>2)) == missing`, despite there existing `eachindex(Dict(1=>2))`. That's intentional but could be confusing. Maybe we need a new name? `iterator_eachindex_type`? `eachindex_iteratortype`?

Maybe we need another trait, `EachIndexable` or something, that includes `Dict`, then `eachindextype` will use that trait instead, and then we combine constraints with `IteratorIndexable` when required.

```julia
iterate(itr::Drop, state...) = _iterate_drop(itr, state..., IteratorIndexable(itr), eachindextype(itr))

function _iterate_drop(it, ::IsIndexable, ::Type{<:AbstractRange})
    [...] # optimized drop
end

function _iterate_drop(it, _, _)
    [...] # default drop
end
```